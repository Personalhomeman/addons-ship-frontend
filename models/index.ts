import { NextContext } from 'next';
import { Store } from 'redux';
import { Request } from 'express';

export type AppVersionProps = {
  appSlug: string;
  versionId: string;
};

export const AppVersionPageTabs = ['details', 'devices', 'test', 'activity'];

export type AppVersionPageQuery = AppVersionProps & {
  isPublic?: string;
  selectedTab?: typeof AppVersionPageTabs[number];
};

export type AppPageQuery = {
  appSlug: string;
};

export type AppVersionList = Array<{
  version: string;
  appVersions: AppVersion[];
}>;

export type AppVersion = {
  id: number;
  version: string;
  platform: string;
  buildNumber: number;
  buildSlug: string;
  lastUpdate: Date;
  description: string;
  whatsNew: string;
  minimumOs: string;
  minimumSdk: string;
  packageName: string;
  certificateExpires: Date;
  bundleId: string;
  size: number;
  supportedDeviceTypes: string[];
  distributionType: string;
  iconUrl: string;
  appName: string;
  promotionalText: string;
  keywords: string;
  reviewNotes: string;
  supportUrl: string;
  marketingUrl: string;
  scheme: string;
  configuration: string;
  publicInstallPageURL: string;
};

export interface PageContext extends NextContext {
  store: Store;
  isServer: boolean;
  req: Request;
}

export type Platform = 'ios' | 'android';
