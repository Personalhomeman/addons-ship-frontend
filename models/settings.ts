export const AppSettingsPageTabs = ['general', 'notifications'];

export type AppSettingsPageQuery = {
  appSlug: string;
  selectedTab?: typeof AppSettingsPageTabs[number];
};

export type IosSettings = {
  artifactExposingWorkflows: string;
  appleDeveloperAccountEmail: string;
  appSku: string;
  appSpecificPassword: string;
  selectedProvProfile: any;
  selectedCertificate: any;
};

export type AndroidSettings = {
  artifactExposingWorkflows: string;
  track: string;
  selectedKeystoreFile: any;
  selectedServiceAccountJsonFile: any;
};

export type ProvProfile = { name: string };
export type Certificate = { name: string };
export type KeystoreFile = { name: string };
export type ServiceAccountJsonFile = { name: string };

export type Settings = {
  iosSettings?: IosSettings;
  androidSettings?: AndroidSettings;
  provProfiles?: ProvProfile[];
  certificates?: Certificate[];
  keystoreFiles?: KeystoreFile[];
  serviceAccountJsonFiles?: ServiceAccountJsonFile[];
  projectType:
    | 'xamarin'
    | 'ios'
    | 'osx'
    | 'macos'
    | 'android'
    | 'cordova'
    | 'ionic'
    | 'react-native'
    | 'flutter'
    | 'other';
};

export type AppContact = {
  id: string;
  email: string;
  isConfirmed: boolean;
  notificationPreferences: {
    newVersion: boolean;
    successfulPublish: boolean;
    failedPublish: boolean;
  };
  isMarkedForDelete?: boolean;
  isMarkedForUpdate?: boolean;
};
