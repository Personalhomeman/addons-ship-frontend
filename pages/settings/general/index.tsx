import { Component } from 'react';
import { connect } from 'react-redux';

import { RootState } from '@/store';

type Props = {};

export class General extends Component<Props> {
  render() {
    return <div>general</div>;
  }
}

const mapStateToProps = ({ testDevices }: RootState) => ({ testDevices });

export default connect(mapStateToProps)(General as any);
