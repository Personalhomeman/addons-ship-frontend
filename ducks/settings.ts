import { Dispatch } from 'redux';
import { createAction, createReducer } from 'deox';

import { ShipAPIService } from '@/services/ship-api';
import { RootState } from '@/store';
import { actionTypeCreator } from '@/utils';
import { Settings } from '@/models/settings';

const $ = actionTypeCreator('SETTINGS');

const _fetchSettings = (appSlug: string) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(fetchSettings.next());

  try {
    const settings = await shipApi.getSettings(appSlug);

    dispatch(fetchSettings.complete(settings));
  } catch (error) {
    dispatch(fetchSettings.error(error));
  }
};

export const fetchSettings = Object.assign(_fetchSettings, {
  next: createAction($`GET_NEXT`),
  complete: createAction($`GET_COMPLETE`, resolve => (settings: Settings) => resolve(settings)),
  error: createAction($`GET_ERROR`, resolve => (error: Error) => resolve(error))
});

const _updateSettings = (settings: Settings) => async (
  dispatch: Dispatch,
  _getState: () => RootState,
  { shipApi }: { shipApi: ShipAPIService }
) => {
  dispatch(updateSettings.next());

  try {
    await shipApi.updateSettings(settings);

    dispatch(updateSettings.complete(settings));
  } catch (error) {
    dispatch(fetchSettings.error(error));
  }
};

export const updateSettings = Object.assign(_updateSettings, {
  next: createAction($`UPDATE_NEXT`),
  complete: createAction($`UPDATE_COMPLETE`, resolve => (settings: Settings) => resolve(settings)),
  error: createAction($`UPDATE_ERROR`, resolve => (error: Error) => resolve(error))
});

const defaultState: SettingsState = {
  projectType: 'other'
};

export type SettingsState = Settings;
export const settings = createReducer(defaultState, handleAction => [
  handleAction(fetchSettings.complete, (_, { payload }) => payload)
]);
