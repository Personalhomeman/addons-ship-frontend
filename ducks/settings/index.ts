import { createReducer } from 'deox';

import { Settings, AppContact } from '@/models/settings';

import fetchSettings from './fetchSettings';
import updateSettings from './updateSettings';
import addAppContact from './addAppContact';
import listAppContacts from './listAppContacts';
import updateAppContact from './updateAppContact';

const defaultState: SettingsState = {
  settings: {
    projectType: 'other'
  },
  appContacts: []
};

export { fetchSettings, updateSettings, addAppContact, listAppContacts, updateAppContact };

export type SettingsState = { settings: Settings; appContacts: AppContact[] };
export default createReducer(defaultState, handleAction => [
  handleAction(fetchSettings.complete, (state, { payload }) => ({ ...state, settings: payload })),
  handleAction(addAppContact.complete, (state, { payload }) => ({
    ...state,
    appContacts: state.appContacts.concat(payload)
  })),
  handleAction(listAppContacts.complete, (state, { payload }) => ({
    ...state,
    appContacts: payload
  })),
  handleAction(updateAppContact.complete, ({ appContacts, ...state }, { payload }) => ({
    ...state,
    appContacts: appContacts.map(appContact => {
      if (appContact.id === payload.id) {
        return payload;
      }

      return appContact;
    })
  }))
]);
