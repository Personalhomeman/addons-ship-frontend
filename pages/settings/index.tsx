import { Component } from 'react';
import Link from 'next/link';
import { Base, Tabs, Tab, Divider, AddonFooter, Flex } from '@bitrise/bitkit';
import startCase from 'lodash/startCase';

import { PageContext } from '@/models';
import { AppSettingsPageQuery, AppSettingsPageTabs } from '@/models/settings';
import { listAppContacts } from '@/ducks/settings';

import General from './general';
import Notifications from './notifications';

export class SettingsPage extends Component<AppSettingsPageQuery> {
  static displayName = 'SettingsPage';
  static defaultProps = {
    selectedTab: 'general'
  };

  static async getInitialProps({ query, store, req, isServer }: PageContext) {
    const { appSlug, selectedTab = AppSettingsPageTabs[0] } = (query as unknown) as AppSettingsPageQuery;

    switch (selectedTab) {
      case 'notifications':
        SettingsPage.displayName = 'NotificationSettings';
        await store.dispatch(listAppContacts(appSlug) as any);
        break;
    }

    return { appSlug, selectedTab };
  }

  tabContent = () => {
    const { selectedTab, appSlug } = this.props;

    switch (selectedTab) {
      case 'general':
        return <General appSlug={appSlug} />;
      case 'notifications':
        return <Notifications appSlug={appSlug} />;
      default:
        return <h1>{selectedTab}</h1>;
    }
  };

  render() {
    const { appSlug, selectedTab } = this.props;

    const tab = (key: string) => (
      <Link as={`/apps/${appSlug}/settings/${key}`} href={`/settings?appSlug=${appSlug}&selectedTab=${key}`}>
        <Tab active={selectedTab === key}>
          <a>{startCase(key)}</a>
        </Tab>
      </Link>
    );

    return (
      <Base>
        <Base>
          <Base maxWidth={960}>
            <Tabs gap="x10">
              {tab('general')}
              {tab('notifications')}
            </Tabs>
          </Base>
          <Divider color="gray-2" direction="horizontal" />
          <Base maxWidth={960}>{this.tabContent()}</Base>
        </Base>
        <Flex paddingVertical="x16">
          <AddonFooter addonName="Ship" />
        </Flex>
      </Base>
    );
  }
}

export default SettingsPage;
