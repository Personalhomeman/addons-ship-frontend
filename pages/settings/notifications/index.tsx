import React from 'react';
import { connect } from 'react-redux';
import isEqual from 'lodash/isEqual';

import { AppContact } from '@/models/settings';
import { RootState } from '@/store';
import { addAppContact } from '@/ducks/settings';

import View from './view';

type Props = {
  appSlug: string;
  appContacts: AppContact[];
  addAppContact: typeof addAppContact;
};

type State = {
  hasModifications: boolean;
  updatedAppContacts: AppContact[];
};

export class NotificationSettings extends React.Component<Props, State> {
  state: State = {
    hasModifications: false,
    updatedAppContacts: this.props.appContacts
  };

  onAddEmail = (email: string) => {
    const { appSlug, addAppContact } = this.props;

    addAppContact(appSlug, email);
  };

  onNotificationPreferenceChanged = (email: string, key: string, value: boolean) => {
    const updatedAppContacts = this.state.updatedAppContacts.map(contact => {
      if (contact.email === email) {
        return {
          ...contact,
          notificationPreferences: {
            ...contact.notificationPreferences,
            [key]: value
          }
        };
      }

      return contact;
    });

    const hasModifications = !isEqual(updatedAppContacts, this.props.appContacts);

    this.setState({ updatedAppContacts, hasModifications });
  };

  onDeleteContact = (email: string) => {
    const updatedAppContacts = this.state.updatedAppContacts.map(contact => {
      if (contact.email === email) {
        return {
          ...contact,
          isMarkedForDelete: true
        };
      }

      return contact;
    });

    this.setState({ updatedAppContacts, hasModifications: true });
  };

  onSave = () => {
    console.log('onSave');
  };

  onCancel = () => {
    const { appContacts } = this.props;

    this.setState({ updatedAppContacts: appContacts, hasModifications: false });
  };

  render() {
    const { updatedAppContacts, hasModifications } = this.state;

    const props = {
      appContacts: updatedAppContacts,
      onAddEmail: this.onAddEmail,
      onNotificationPreferenceChanged: this.onNotificationPreferenceChanged,
      onDeleteContact: this.onDeleteContact,
      onSave: hasModifications ? this.onSave : undefined,
      onCancel: this.onCancel
    };

    return <View {...props} />;
  }
}

const mapStateToProps = ({ settings: { appContacts } }: RootState) => ({ appContacts }),
  mapDispatchToProps = { addAppContact };

export default connect(
  mapStateToProps,
  mapDispatchToProps
  // @ts-ignore
)(NotificationSettings);
