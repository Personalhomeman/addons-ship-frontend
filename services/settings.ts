import { AppVersion } from '@/models';
import { IosSettings, AndroidSettings } from '@/models/settings';

export class SettingsService {
  constructor() {}

  isComplete = (
    appVersion: AppVersion,
    { iosSettings, androidSettings }: { iosSettings: IosSettings; androidSettings: AndroidSettings }
  ) => {
    switch (appVersion.platform) {
      case 'ios': {
        return ![
          'appleDeveloperAccountEmail',
          'appSku',
          'appSpecificPassword',
          'selectedAppStoreProvisioningProfile',
          'selectedCodeSigningIdentity'
        ].find(key => !iosSettings[key]);
      }
      case 'android': {
        return !['track', 'selectedKeystoreFile', 'selectedServiceAccount'].find(key => !androidSettings[key]);
      }
    }

    return true;
  };
}

export default new SettingsService();
