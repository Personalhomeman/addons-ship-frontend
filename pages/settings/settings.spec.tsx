jest.mock('@/ducks/settings');

import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { listAppContacts } from '@/ducks/settings';
import { AppSettingsPageTabs } from '@/models/settings';

import { SettingsPage } from './';

describe('Settings', () => {
  describe('renders tabs correctly', () => {
    [...AppSettingsPageTabs, 'other'].forEach(tab => {
      test(tab, () => {
        const tree = shallowToJson(shallow(<SettingsPage selectedTab={tab} appSlug="some-app" />));
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe('dispatches the proper actions', () => {
    it('dispatches listAppContacts', async () => {
      await SettingsPage.getInitialProps({
        query: { selectedTab: 'notifications' },
        store: { dispatch: jest.fn() }
      } as any);

      expect(listAppContacts).toHaveBeenCalled();
    });
  });
});
