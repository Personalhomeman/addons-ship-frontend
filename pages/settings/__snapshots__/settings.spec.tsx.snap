// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings renders tabs correctly general 1`] = `
<Base>
  <Base>
    <Base
      maxWidth={960}
    >
      <Tabs
        gap="x10"
      >
        <Link
          as="/apps/some-app/settings/general"
          href="/settings?appSlug=some-app&selectedTab=general"
        >
          <Tab
            active={true}
          >
            <a>
              General
            </a>
          </Tab>
        </Link>
        <Link
          as="/apps/some-app/settings/notifications"
          href="/settings?appSlug=some-app&selectedTab=notifications"
        >
          <Tab
            active={false}
          >
            <a>
              Notifications
            </a>
          </Tab>
        </Link>
      </Tabs>
    </Base>
    <Divider
      color="gray-2"
      direction="horizontal"
      width="0.125rem"
    />
    <Base
      maxWidth={960}
    >
      <Connect(GeneralSettings)
        appSlug="some-app"
      />
    </Base>
  </Base>
  <Flex
    paddingVertical="x16"
  >
    <AddonFooter
      addonName="Ship"
    />
  </Flex>
</Base>
`;

exports[`Settings renders tabs correctly notifications 1`] = `
<Base>
  <Base>
    <Base
      maxWidth={960}
    >
      <Tabs
        gap="x10"
      >
        <Link
          as="/apps/some-app/settings/general"
          href="/settings?appSlug=some-app&selectedTab=general"
        >
          <Tab
            active={false}
          >
            <a>
              General
            </a>
          </Tab>
        </Link>
        <Link
          as="/apps/some-app/settings/notifications"
          href="/settings?appSlug=some-app&selectedTab=notifications"
        >
          <Tab
            active={true}
          >
            <a>
              Notifications
            </a>
          </Tab>
        </Link>
      </Tabs>
    </Base>
    <Divider
      color="gray-2"
      direction="horizontal"
      width="0.125rem"
    />
    <Base
      maxWidth={960}
    >
      <Connect(NotificationSettings)
        appSlug="some-app"
      />
    </Base>
  </Base>
  <Flex
    paddingVertical="x16"
  >
    <AddonFooter
      addonName="Ship"
    />
  </Flex>
</Base>
`;

exports[`Settings renders tabs correctly other 1`] = `
<Base>
  <Base>
    <Base
      maxWidth={960}
    >
      <Tabs
        gap="x10"
      >
        <Link
          as="/apps/some-app/settings/general"
          href="/settings?appSlug=some-app&selectedTab=general"
        >
          <Tab
            active={false}
          >
            <a>
              General
            </a>
          </Tab>
        </Link>
        <Link
          as="/apps/some-app/settings/notifications"
          href="/settings?appSlug=some-app&selectedTab=notifications"
        >
          <Tab
            active={false}
          >
            <a>
              Notifications
            </a>
          </Tab>
        </Link>
      </Tabs>
    </Base>
    <Divider
      color="gray-2"
      direction="horizontal"
      width="0.125rem"
    />
    <Base
      maxWidth={960}
    >
      <h1>
        other
      </h1>
    </Base>
  </Base>
  <Flex
    paddingVertical="x16"
  >
    <AddonFooter
      addonName="Ship"
    />
  </Flex>
</Base>
`;
